### PR TITLE
8254999: Move G1RemSetSamplingTask to more appropriate location

### PR DESCRIPTION
**Pull request by kstefanj**
Please review this change that moves the `Remembered Set Sampling Task` from `g1ServiceThread.cpp` into `g1RemSet.cpp`. This is basically just a move but with a small edition that the task will reschedule itself instead of running if at start of execution the last GC happend more recently than the value of `G1ConcRefinementServiceIntervalMillis`.

This change does not cover moving the timing measurements from the service thread into the task, that will be covered by: 
[JDK-8252645](https://bugs.openjdk.java.net/browse/JDK-8252645)

Testing
Tier1-3 + local verification that the tast is scheduled as expected.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) |    |  ✔️ (9/9 passed) | ❌ (1/9 failed) |

**Failed test tasks**
- [Linux x86 (build debug)](https://github.com/kstefanj/jdk/runs/1447750536)
- [Linux x86 (build release)](https://github.com/kstefanj/jdk/runs/1447750516)
- [macOS x64 (hs/tier1 runtime)](https://github.com/kstefanj/jdk/runs/1447913827)

### Issue
 * [JDK-8254999](https://bugs.openjdk.java.net/browse/JDK-8254999): Move G1RemSetSamplingTask to more appropriate location

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to e6b7f959ce621325e80e52fd1fd24509511ac3c4

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1395/head:pull/1395`
`$ git checkout pull/1395`

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 1]
```diff
@@ -488,17 +490,122 @@ G1RemSet::G1RemSet(G1CollectedHeap* g1h,
...
+    // Reschedule if a GC happened too recently.
+    if (should_reschedule()) {
+      // Calculate the delay given the last GC and the interval.
+      schedule(reschedule_delay_ms());
+      return;
+    }
+
+    // Do the actual sampling.
+    sample_young_list_rs_length(&sts);
+    schedule(G1ConcRefinementServiceIntervalMillis);
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:599]: I am good with the new location of the code but I would prefer if this change would be limited to the move and not change the rescheduling logic.
If it's not too much of a bother please split this out into a second CR.

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1RemSet.cpp (hunk 1 above)
Comment (kstefanj) [reply src/hotspot/share/gc/g1/g1RemSet.cpp]: Thanks for looking at this Thomas. I removed the scheduling logic and filed: [JDK-8252645](https://bugs.openjdk.java.net/browse/JDK-8252645)

**Review by tschatzl — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 2]
```diff
@@ -488,17 +490,97 @@ G1RemSet::G1RemSet(G1CollectedHeap* g1h,
...
   _ct(ct),
   _g1p(_g1h->policy()),
-  _hot_card_cache(hot_card_cache) {
+  _hot_card_cache(hot_card_cache),
+  _sampling_task(NULL) {
 }
 
 G1RemSet::~G1RemSet() {
   delete _scan_state;
+  if (_sampling_task != NULL) {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:499]: Could you add a comment to explain when `_sampling_task` would be null?

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1RemSet.cpp (hunk 2 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:499]: Actually the check can be removed because the operator implementation handles NULL (by ignoring it).

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:499]: Even better. Thanks for the suggestion.

Comment (kstefanj) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:499]: Cool, then I will just remove the NULL-check.

**Review by tschatzl — APPROVED** [review]
Still lgtm

**Comment by kstefanj** [comment]
Thanks @albertnetymk and @tschatzl for the reviews.
/integrate

**Pull request closed without merging**